god-mode.el

```diff
@@ -461,6 +462,78 @@ be ignored by `god-execute-with-current-bindings'."
                        universal-argument
                        universal-argument-more)))
 
+(defcustom god-translate-key-for-description
+  t
+  "Whether to use `god-mode-describe-key' when god-mode is enabled."
+  :group 'god
+  :type 'boolean)
+
+(defvar god-latest-described-command nil
+  "The latest command recorded by `god-mode-describe-key'.")
+
+(defun god-mode--help-fn-describe-function (_arg)
+  "Insert information about `god-mode' key-bindings for the described function.
+The argument _ARG is ignored: it's only needed because all hooks in
+`help-fns-describe-function-functions' take the function-name as argument.
+But in our case it's redundant"
+  (insert
+   (format-message "\n  %s\n  %s%s\n  %s%s%s\n\n"
+		   "(`god-local-mode' is enabled. "
+		   " The given key-sequence: "
+		   (god-mode-get-described-key-seq)
+		   " corresponds to this key-binding: "
+		   (help--key-description-fontified
+		    god-latest-described-command)
+		   ")")))
+
+(defun god-mode-get-described-key-seq ()
+  "Return the keys that were pressed after `god-mode-describe-key' was called."
+  (let* ((latest-keys (recent-keys 'include-cmds))
+	 (latest-describe-key-index
+	  (cl-position '(nil . god-mode-self-insert)
+		       latest-keys :from-end t :test #'equal))
+	 (start-index (+ 3 latest-describe-key-index))
+	 (key-sequence (key-description (seq-subseq latest-keys start-index))))
+    (propertize key-sequence
+		'font-lock-face 'help-key-binding
+                'face 'help-key-binding)))
+
+(defun god-mode-describe-key ()
+  "Describe a key-sequence as interpreted by `god-mode'.
+Prompt for a key sequence, use `god-mode-lookup-key-sequence' to translate it
+into the appropriate command, and use `describe-function' to describe it.
+Only applied when `god-translate-key-for-description' is t:
+when nil, `describe-key' is called instead"
+  (interactive)
+  (if (not god-translate-key-for-description)
+      (call-interactively #'describe-key)
+    (progn
+      (message "Describe the following god-mode key: ")
+      (advice-add #'god-mode-lookup-command :filter-args
+		  (lambda (key-string)
+		    (setq god-latest-described-command key-string)))
+      (let ((command
+	     ;; if the key is not recognized by god-mode,
+	     ;; we will pass it to the regular `describe-key'
+	     (condition-case err
+		 (god-mode-lookup-key-sequence)
+	       (wrong-type-argument
+		;; due to  how errors are passed,
+		;; we do not have enough information
+		;; to pass menu items
+		(if (not (equal (cddr err) '((menu-bar))))
+		  (describe-key (vector (caddr err))))))))
+	(if command
+	    (progn
+	      (add-hook 'help-fns-describe-function-functions
+			#'god-mode--help-fn-describe-function)
```

Comment:
Since you're using hooks, is it possible to add the hook when `god-local-mode` is enabled, and remove it when the mode is disabled?

Reply:
I don't really think this is doable without adding a lot of complexity.
This because of two reasons:

1. One problem with adding advices when `god-local-mode` is enabled, and removing them when it is disabled, is that we may have different "instances" of `god-local-mode` across different buffers. This creates a problem if, for example, we have a workflow along these lines:

- enable `god-local-mode` in buffer A. The hooks/advices for `god-mode-describe-key` are added
- switch to another buffer B, enable `god-local-mode`. The hooks/advices are "added" but since they are already there, this makes no difference
- disable `god-local-mode` in buffer B. The hooks are removed
- switch back to buffer A. Now the hooks/advices necessary for `god-mode-describe-key` have been removed, and the function doesn't work properly anymore.

I can't think of simple ways to fix this.

2. Another problem with this specific function, is that it is added to the describe _function_ hook, so if we simply add it as a hook when we enable `god-mode,` any call to `describe-function` would mistakenly try to lookup the inputted keys and their god-mode translation. The way to fix this could be to make `god-mode--help-fn-describe-function` discriminate whether it is being called from `god-mode-describe-key` or from a "regular" call to `describe-function`, and not do anything in the latter case. But I can't think of a way to do this without introducing a new variable: `god-mode-describe-key` could set a boolean to True when called and then set it back to False after the call to `describe-function`. To me this approach seems more complex and unnecessarily clunky than the current one.

Reply:
Good point. Let's leave this as is.